File: curriculum/algos/merge_sort.cpp
// Generic stable merge sort over random-access iterators.
#include <algorithm>
#include <functional>
#include <iterator>
#include <vector>
template<typename Iterator, typename Compare>
void merge_sort_impl(Iterator first, Iterator last, Compare compare) {
auto length = std::distance(first, last);
if (length < 2) return;
Iterator middle = std::next(first, length / 2);
merge_sort_impl(first, middle, compare);
merge_sort_impl(middle, last, compare);
using Value = typename std::iterator_traits<Iterator>::value_type;
std::vector<Value> buffer;
buffer.reserve(length);
Iterator left = first;
Iterator right = middle;
while (left != middle && right != last) {
if (compare(*right, *left)) buffer.push_back(*right++);
else buffer.push_back(*left++);
}
while (left != middle) buffer.push_back(*left++);
while (right != last) buffer.push_back(*right++);
std::move(buffer.begin(), buffer.end(), first);
}
template<typename Iterator, typename Compare = std::less<>>
void merge_sort(Iterator first, Iterator last, Compare compare = {}) {
merge_sort_impl(first, last, compare);
}
// Explicit recursive merge-sort variant.
template<typename Iterator, typename Compare>
void recursive_merge_sort_impl(Iterator first, Iterator last, Compare compare) {
auto length = std::distance(first, last);
if (length < 2) return;
Iterator middle = std::next(first, length / 2);
recursive_merge_sort_impl(first, middle, compare);
recursive_merge_sort_impl(middle, last, compare);
using Value = typename std::iterator_traits<Iterator>::value_type;
std::vector<Value> buffer;
buffer.reserve(length);
Iterator left = first;
Iterator right = middle;
while (left != middle && right != last) {
if (compare(*right, *left)) buffer.push_back(*right++);
else buffer.push_back(*left++);
}
while (left != middle) buffer.push_back(*left++);
while (right != last) buffer.push_back(*right++);
std::move(buffer.begin(), buffer.end(), first);
}
template<typename Iterator, typename Compare = std::less<>>
void recursive_merge_sort(Iterator first, Iterator last, Compare compare = {}) {
recursive_merge_sort_impl(first, last, compare);
}

File: curriculum/algos/MergeSort.java
final class MergeSort {
static void mergeSort(int[] values) {
int[] buffer = new int[values.length];
for (int width = 1; width < values.length; width *= 2) {
for (int first = 0; first < values.length; first += width * 2) {
int middle = Math.min(first + width, values.length);
int last = Math.min(first + width * 2, values.length);
merge(values, buffer, first, middle, last);
}
}
}
static void recursiveMergeSort(int[] values) {
int[] buffer = new int[values.length];
recursiveMergeSort(values, buffer, 0, values.length);
}
private static void recursiveMergeSort(int[] values, int[] buffer, int first, int last) {
if (last - first < 2) return;
int middle = first + (last - first) / 2;
recursiveMergeSort(values, buffer, first, middle);
recursiveMergeSort(values, buffer, middle, last);
merge(values, buffer, first, middle, last);
}
private static void merge(int[] values, int[] buffer, int first, int middle, int last) {
int left = first;
int right = middle;
int output = first;
while (left < middle && right < last)
buffer[output++] = values[right] < values[left] ? values[right++] : values[left++];
while (left < middle) buffer[output++] = values[left++];
while (right < last) buffer[output++] = values[right++];
for (int index = first; index < last; ++index) values[index] = buffer[index];
}
}
